attachment_preview/models/ir_attachment.py

```diff
@@ -0,0 +1,83 @@
+# Copyright 2014 Therp BV (<http://therp.nl>)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+import collections
+import logging
+import mimetypes
+import os.path
+
+from odoo import api, models
+
+_logger = logging.getLogger(__name__)
+
+
+class IrAttachment(models.Model):
+    _inherit = "ir.attachment"
+
+    @api.model
+    def get_binary_extension(self, model, ids, binary_field, filename_field=None):
+        # import pdb
+        # pdb.set_trace()
+        result = {}
+        ids_to_browse = ids if isinstance(ids, collections.Iterable) else [ids]
+
+        # First pass: load fields in bin_size mode to avoid loading big files
+        #  unnecessarily.
+        if filename_field:
+            for this in (
+                self.env[model].with_context(
+                    bin_size=True).browse(ids_to_browse)
+            ):
+                if not this.id:
+                    result[this.id] = False
+                    continue
+                extension = ""
+                if this[filename_field]:
+                    filename, extension = os.path.splitext(
+                        this[filename_field])
+                if this[binary_field] and extension:
+                    result[this.id] = extension
+                    _logger.debug(
+                        "Got extension %s from filename %s",
+                        extension,
+                        this[filename_field],
+                    )
+        # Second pass for all attachments which have to be loaded fully
+        #  to get the extension from the content
+        ids_to_browse = [_id for _id in ids_to_browse if _id not in result]
+        for this in self.env[model].with_context(bin_size=True).browse(ids_to_browse):
+            if not this[binary_field]:
+                result[this.id] = False
+                continue
+            try:
```

Comment:
Move the import of magic to the top of the file, after setting _logger, like so:
```
try:
    import magic
except ImportError:
    _logger.info(_("Library magic is not available, will guess mime types."))
    magic = None
```
Also add import of `_` (translations) from odoo to line importin gapi and models.
Add translation to all log messages.

In the get_binary_extention method test if magic == None, and in that case use the mimetypes.guess_type() method, instead of doing this in an exception handler. Better by the way to break up this large method in smaller pieces.